explainaboard/loaders/loader.py

```diff
@@ -87,78 +128,66 @@ def user_defined_metadata_configs(self) -> dict:
             )
         return self._user_defined_metadata_configs
 
-    def _parse_user_defined_features_configs(self) -> dict:
-        if self._file_type == FileType.json:
-            raw_data = self.file_loaders[FileType.json].load_raw(
-                self._data, self._source
+    def _parse_user_defined_fields(self) -> tuple[dict[str, CustomFeature], dict]:
+        """custom features and metadata can only be defined in the output file and it
```

Comment:
This is something that might need discussion. There are certainly instances where we would want the user-defined fields within the dataset. For example, the MetaphorQA dataset, for which we're planning to build a leaderboard, has annotations over the type of questions included in the dataset itself.

On the other hand, ExplainaBoard users might want to define their own features to use in analyses, in which case they'd want to upload them either as part of the system output, or as a separate file.

I'm not sure if this is important enough to block this PR, but it's worth thinking about what our future strategy will be.

Reply:
I see. I think putting `user_defined_features_configs` and `user_defined_metadata_configs` in the dataset or output file can be a little awkward in some cases because it would mean that the file format can be different depending on whether user-defined configs exist. It also makes the JSON format a special case because user-defined fields can only be specified with this file type.

Also, instead of passing in the task name explicitly, the SDK now supports passing in the task names in the metadata config. But that means that we need to parse the file before knowing which loader we need to parse it with (loaders are task-dependent). Currently, we use dummy task types to parse the file once to get the metadata and then create an actually task-specific loader to parse the file a second time. (this is in **explainaboard_main.py**).

I think maybe it would make sense to move user-defined configs to a separate file but this would require our users to provide one more file. If we move these configs to a separate file, I think we can allow configs and metadata to come from two sources: a config file (maybe yaml or json) and the metadata argument passed to `get_loader` or `get_processor` in the code. We will merge the two to get the full metadata. I think part of this is already implemented in **explainaboard_main.py** as a special case where the task names are extracted from the output file and added to the metadata dict. We can maybe support this feature for all general cases.